### ISSUE TITLE
Attribute Names/Values are unexpected when building an update expression.

### ISSUE THREAD

**Issue by MinisculeGirraffe** [neoeinstein/modyne#19]
Firstly, this is such a cool library and I don't think i'll ever go back to the old ways of interacting with DynamoDB in Rust again. My coworkers will be forever annoyed as I try and shim this into our codebase. 

This library pairs VERY well with [async-graphql](https://github.com/async-graphql/async-graphql). Combining these solves lots of the use cases of AWS AppSync and IMO is way nicer to work with/cheaper. 

I'm aware that it's possible to use [update in the EntityExt](https://docs.rs/modyne/latest/modyne/trait.EntityExt.html#method.update) to apply just specific updates to the entity.

Example of a toy project i'm working with. 

```Rust
#[derive(Debug, InputObject)]
pub struct DocumentKey {
    pub document_id: String,
}

#[derive(
    Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize, Enum,
)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DocumentVisibility {
    Public,
    Private,
}
#[derive(Clone, Debug, modyne::EntityDef, serde::Serialize, serde::Deserialize, SimpleObject)]
pub struct Document {
    pub id: String,
    pub title: String,
    pub visibility: DocumentVisibility,
    pub created_at: DateRFC3339,
}

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, InputObject)]
pub struct UpdateDocumentInput {
    pub id: String,
    pub title: Option<String>,
    pub visibility: Option<String>,
}
```

I find myself writing a lot of boiler plate for updating values, which basically boils down to the 

```Rust
    pub async fn update_document(&self, input: UpdateDocumentInput) -> Result<(), Error> {
        let key = DocumentKey {
            document_id: input.document_id,
        };

        let mut expr = expr::Update::new("");

        if let Some(title) = input.title {
            expr.expression.push_str("SET #title = :title");
            expr = expr.name("#title", "title");
            expr = expr.value(":title", title)
        }

        if let Some(visibility) = input.visibility {
            expr.expression.push_str("SET #visibility = :visibility");
            expr = expr.name("#visibility", "visibility)
            expr = expr.value(":visibility", visibility);
        }

        DocumentDynamo::update(key)
            .expression(expr)
            .execute(&self.db)
            .await?;

        Ok(())
    }
```

This caused an error because the [Update](https://docs.rs/modyne/latest/modyne/expr/struct.Update.html) struct will append strings to the values in the expression to ensure that it's unique to the operation type. 

For my function to work I had to actually write this line when directly manipulating the update expression to keep it in the format of the string manipulation. 

```Rust
        if let Some(title) = input.title {
            expr.expression.push_str("SET #upd_title = :upd_title");
            expr = expr.name("#title", "title");
            expr = expr.value(":title", title)
        }

        if let Some(visibility) = input.visibility {
            expr.expression.push_str("SET #upd_visibility = :upd_visibility");
            expr = expr.name("#visibility", "visibility");
            expr = expr.value(":visibility", visibility);
        }
```

This was kind of an unexpected foot gun because the fields in the struct were public, so I assumed there wouldn't be any special business logic involved in manipulating them.

**Comment by MinisculeGirraffe** [comment]
I feel like this is actually an extension of #12

**Comment by neoeinstein** [comment]
> This was kind of an unexpected foot gun because the fields in the struct were public, so I assumed there wouldn't be any special business logic involved in manipulating them.

Yeah, that's on me. I would probably update it to be private, and allow controlled manipulation rather than direct mutability.

There's definitely some boilerplate, and having something like the serde "patch" pattern you mentioned before would work for several cases, but it leaves out some important cases, which is why I didn't lean into it at first. I might be able to extend the code generation to create a "patch" builder struct for this case, but updates for sets based on functions or prior values, increment, and add/remove from sets would still need the lower level of control intended by `expr::Update`.

My alternate recommendation, building up the expression template and _then_ creating the expression would also be boilerplate-y and frankly ugly, requiring the `if let`s multiple times, and would certainly be a reason for preferring a patch struct instead.

So, I think I would want to provide an implementation (I'd mark it as deprecated for now rather than do a major version bump), and then add a simple function that would allow for _building_ of an expression string as you are intending to do here.

Adding a patch struct or extension trait that's customized for a given entity would be a secondary task more in line with #12, and would likely build upon that first step, and would be in an "as I or some kind soul has time" timeframe.

Does that sound reasonable?

**Comment by MinisculeGirraffe** [comment]
Thanks for the reply. And yeah agreed. I already added the method in a local brach I had. 

I also wrote the proc macro for #12 actually too.  Example of the generated code:

``` rust
 impl Into<::modyne::expr::Update> for DocumentDynamoPatch {
        fn into(self) -> ::modyne::expr::Update {
            let mut expr = expr::Update::new("");
            if let Some(title) = &self.title {
                expr = expr.add_expression("SET #title = :title");
                expr = expr.name("#title", "title");
                expr = expr.value(":title", title);
            }
            if let Some(visibility) = &self.visibility {
                expr = expr.add_expression("SET #visibility = :visibility");
                expr = expr.name("#visibility", "visibility");
                expr = expr.value(":visibility", visibility);
            }
            expr
        }
    }
```

I'm going to try and get a PR through today if that's cool.

**Comment by neoeinstein** [comment]
> I'm going to try and get a PR through today if that's cool.

Great. I look forward to seeing it posted!
